### PR TITLE
Implement Frame Management

### LINKED ISSUE #163: Implement Frame Management
https://github.com/GoogleChrome/puppeteer/blob/v1.0.0/test/test.js#L1641

### PR DESCRIPTION
**Pull request by kblok**
closes #163

### CODE REVIEW

File: lib/PuppeteerSharp.Tests/Frame/FrameManagementTests.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            Assert.Equal(4, detachedFrames.Count);
+            Assert.Single(navigatedFrames);
+        }
+
+        [Fact]
+        public async Task ShouldReportFrameName()
+        {
+            await FrameUtils.AttachFrame(Page, "theFrameId", TestConstants.EmptyPage);
+            await Page.EvaluateFunctionAsync(@"url =>
+            {
```
Comment (Meir017) [inline lib/PuppeteerSharp.Tests/Frame/FrameManagementTests.cs]: move this to the previous line (js script...)

File: lib/PuppeteerSharp.Tests/TestConstants.cs [hunk 2]
```diff
@@ -17,7 +17,13 @@ public static class TestConstants
 
         public static readonly DeviceDescriptor IPhone = DeviceDescriptors.Get(DeviceDescriptorName.IPhone6);
         public static readonly DeviceDescriptor IPhone6Landscape = DeviceDescriptors.Get(DeviceDescriptorName.IPhone6Landscape);
-        
+
+        public static readonly string NestedFramesDumpResult = @"http://127.0.0.1:<PORT>/frames/nested-frames.html
```
Comment (Meir017) [inline lib/PuppeteerSharp.Tests/TestConstants.cs]: maybe create a class `GoldenUtils` that will contain all of the golden files?

Comment (kblok) [reply lib/PuppeteerSharp.Tests/TestConstants.cs]: I'm not so sure yet. That golden guy is used to compare files. They used that tool to compare this dump and I think it makes no sense for them to do that.

File: lib/PuppeteerSharp.Tests/Frame/WaitForSelectorTests.cs [hunk 3]
```diff
@@ -73,12 +73,12 @@ public class WaitForSelectorTests : PuppeteerPageBaseTest
...
         {
-            await FrameUtils.AttachFrame(Page, "frame1", TestConstants.EmptyPage);
-            await FrameUtils.AttachFrame(Page, "frame2", TestConstants.EmptyPage);
+            await FrameUtils.AttachFrameAsync(Page, "frame1", TestConstants.EmptyPage);
+            await FrameUtils.AttachFrameAsync(Page, "frame2", TestConstants.EmptyPage);
             var frame1 = Page.Frames.ElementAt(1);
             var frame2 = Page.Frames.ElementAt(2);
             var added = false;
-            frame2.WaitForSelectorAsync("div").ContinueWith(_ => added = true);
+            var selectorTask = frame2.WaitForSelectorAsync("div").ContinueWith(_ => added = true);
```
Comment (Meir017) [inline lib/PuppeteerSharp.Tests/Frame/WaitForSelectorTests.cs]: I guess this to ignore the warning 😏

File: lib/PuppeteerSharp/Page.cs [hunk 4]
```diff
@@ -101,7 +96,7 @@ private Page(Session client, Target target, FrameTree frameTree, bool ignoreHTTP
         public event EventHandler<RequestEventArgs> RequestFailed;
 
         public Frame MainFrame => _frameManager.MainFrame;
-        public IEnumerable<Frame> Frames => _frameManager.Frames.Values;
+        public Frame[] Frames => _frameManager.Frames.Values.ToArray();
```
Comment (Meir017) [inline lib/PuppeteerSharp/Page.cs]: docs for these properties?

**Comment by Meir017** [comment]
looks good, just missing some docs

File: lib/PuppeteerSharp.Tests/Frame/WaitForSelectorTests.cs (hunk 3 above)
Comment (kblok) [reply lib/PuppeteerSharp.Tests/Frame/WaitForSelectorTests.cs]: Yup :)

File: lib/PuppeteerSharp/Page.cs (hunk 4 above)
Comment (kblok) [reply lib/PuppeteerSharp/Page.cs]: Done @Meir017

**Pull request merged**